scalding-beam/src/main/scala/com/twitter/scalding/beam_backend/BeamOp.scala

```diff
@@ -59,6 +61,25 @@ object BeamOp extends Serializable {
   )(implicit ordK: Ordering[K], kryoCoder: KryoCoder): PCollection[KV[K, java.lang.Iterable[U]]] = {
     reduceFn match {
       case ComposedMapGroup(f, g) => planMapGroup(planMapGroup(pcoll, f), g)
+      case EmptyGuard(MapValueStream(SumAll(pqm: PriorityQueueMonoid[V]))) =>
+        pcoll.apply(MapElements.via(
+          new SimpleFunction[KV[K, java.lang.Iterable[V]], KV[K, java.lang.Iterable[U]]]() {
+            override def apply(input: KV[K, lang.Iterable[V]]): KV[K, java.lang.Iterable[U]] = {
+              // We are not using plus method defined in PriorityQueueMonoid as it is mutating
+              // input Priority Queues. We create a new PQ from the individual ones.
+              // We didn't use Top PTransformation in beam as it is not needed, also
+              // we cannot access `max` defined in PQ monoid.
+              val flattenedValues = input.getValue.asScala.flatMap { value =>
```

Comment:
I think this is going to materialize everything in memory. I think you should do `input.getValue.asScala.toStream.flatMap`... to make sure this is a lazy Iterable.

Reply:
`input.getValue.asScala` gives an Iterator only and `flatmap` also works lazily. Which call would materialize it?

Reply:
`getValue` returns the `Iterable[_]` no? so, asScala on that is going to give you an `Iterable` if I'm not mistaken. You could call `.iterator` on that before you call .asScala and that will certainly be lazy, which is better than the `toStream` since we would rather not materialize the whole set.

Reply:
actually... `build` requires an `Iterable` and since `flattenedValues` is a `val` it will pin things in memory.

To take this suggestion, I think you need `@inline def flattenedValues = input.getValue.asScala.toStream.flatMap...`

Reply:
I think I understand all the pieces but would really appreciate if you could clarify how to make it properly lazy.

In the first part are you referring to `Iterable` keeping references to all elements, as opposed to `Iterator` which wouldn't?

Then when you mention using inline, is it to avoid a closure over that variable keeping the reference alive for longer or landing in more expensive generation? Or more about escape analysis or similar?

Reply:
In scala, most Iterable methods (like flatMap) are not lazy and will materialize the entire result in many cases (basically the only exception is Stream or LazyList in scala 2.13).

An Iterator generally drops references to the items it has iterated past, although you could imagine making one that doesn't. Consider a List iterator:
```scala
class ListIterator[A](lst: List[A]) extends Iterator[A] {
  private[this] var current: List[A] = lst
  def hasNext: Boolean = current.nonEmpty
  def next(): A = {
    val result = current.head
    current = current.tail
    result
  }
}
```
The way the scala compiler works, `lst` is only referenced by the constructor, so you won't keep a reference to that beyond the constructor which just uses it to initialize.

Finally, the `@inline` is just a hint, but the `def` means there is nothing keeping the reference around so when the GC runs it could colllect anything not pointed to by the Stream at that moment.

BUT, all of this is a bit moot because looking inside `PriorityQueueMonoid` we see that `build` traverses the Iterable twice... so if it is a large `Stream` you will materialize the whole thing...

I think the best idea would be just subclass PriorityQueueMonoid so you can access the max size and the ordering, and use that call Top.of.